tooling/ssa_fuzzer/src/builder.rs

```diff
@@ -0,0 +1,309 @@
+#![allow(dead_code)]
+use crate::compiler::compile;
+use crate::config::NUMBER_OF_VARIABLES_INITIAL;
+use crate::helpers;
+use acvm::FieldElement;
+use noirc_driver::{CompileError, CompileOptions, CompiledProgram};
+use noirc_evaluator::ssa::function_builder::FunctionBuilder;
+use noirc_evaluator::ssa::ir::basic_block::BasicBlockId;
+use noirc_evaluator::ssa::ir::function::{Function, RuntimeType};
+use noirc_evaluator::ssa::ir::instruction::BinaryOp;
+use noirc_evaluator::ssa::ir::map::Id;
+use noirc_evaluator::ssa::ir::types::{NumericType, Type};
+use noirc_evaluator::ssa::ir::value::Value;
+use noirc_frontend::monomorphization::ast::InlineType as FrontendInlineType;
+use std::sync::Arc;
+
+/// Builder for generating fuzzed SSA functions
+/// Contains a FunctionBuilder and tracks the current numeric type being used
+pub struct FuzzerBuilder {
+    builder: FunctionBuilder,
+    numeric_type: NumericType,
+    type_: Type,
+}
+
+impl FuzzerBuilder {
+    /// Creates a new FuzzerBuilder in ACIR context
+    pub fn new_acir() -> Self {
+        let main_id: Id<Function> = Id::new(0);
+        let mut builder = FunctionBuilder::new("main".into(), main_id);
+        builder.set_runtime(RuntimeType::Acir(FrontendInlineType::default()));
+        Self {
+            builder,
+            numeric_type: NumericType::NativeField,
+            type_: Type::Numeric(NumericType::NativeField),
+        }
+    }
+
+    /// Creates a new FuzzerBuilder in Brillig context
+    pub fn new_brillig() -> Self {
+        let main_id: Id<Function> = Id::new(0);
+        let mut builder = FunctionBuilder::new("main".into(), main_id);
+        builder.set_runtime(RuntimeType::Brillig(FrontendInlineType::default()));
+        Self {
+            builder,
+            numeric_type: NumericType::NativeField,
+            type_: Type::Numeric(NumericType::NativeField),
+        }
+    }
+
+    /// Compiles the built function into a CompiledProgram, to run it with nargo execute
+    pub fn compile(self) -> Result<CompiledProgram, CompileError> {
+        compile(self.builder, &CompileOptions::default())
+    }
+
+    /// Inserts initial variables of the given type into the function
+    pub fn insert_variables(&mut self, variable_type: Type) {
+        for _ in 0..NUMBER_OF_VARIABLES_INITIAL {
+            self.builder.add_parameter(variable_type.clone());
+        }
+        self.type_ = variable_type.clone();
+        match variable_type {
+            Type::Numeric(numeric_type) => {
+                self.numeric_type = numeric_type;
+            }
+            _ => {
+                panic!("Unsupported variable type");
```

Comment:
Maybe the check could be done first, no?

Reply:
It looks like you could have called `set_type`, although I still think it would be worth considering setting the type in the constructor, and verifying that the variables being inserted here are compatible, rather than allow changing the type.